### DISCUSSION TITLE
Feature Request: Kindly add LLVM `clang-cl` to the list of supported compilers

### DISCUSSION THREAD

**Discussion by emmenlau** [duckdb/duckdb#6487] [Ideas]
When developing for Microsoft Windows, the LLVM clang-cl compiler frontend for Visual Studio can be a very nice alternative to using the MSVC compilers directly. Benefits of clang-cl are faster code (for us, typically around 5-15% faster, depending on the project and application), and better standardization when compared to Un*x development (it behaves properly like a Microsoft compiler but accepts a lot of typical Un*x extensions, styles, logging, etc. The LLVM compiler also supports more fine-grained control when optimizing for specific CPU instructions sets, and for us, it accepts some very complex templated c++ that Visual Studio compilers will ICE on.

The compiler is fully supported by Microsoft: https://learn.microsoft.com/en-us/cpp/build/clang-support-msbuild?view=msvc-170

duckdb 0.7.1 can build almost out of the box with `clang-cl`, except for some very minor changes.

All that is needed to support this compiler is:
 - Download a recent version like https://github.com/llvm/llvm-project/releases/download/llvmorg-15.0.7/LLVM-15.0.7-win64.exe
 - Install (basically just extract the archive)
 - Add the extracted `bin` directory to PATH
 - Set environment variables `CC=clang-cl` and `CXX=clang-cl`
 - Run cmake like before...

**Comment by Tishj** [comment]
I'm not entirely sure what you mean by "add to supported compilers"
Are you asking that we add a CI run for it?

**Comment by emmenlau** [comment]
Hey @Tishj!
Basically what I mean is that duckdb would "officially support clang-cl", which in my eyes would mean that the code builds, but also that clang-cl specific differences are addressed when needed.

That could probably mean some extra effort in coding. For example, clang-cl may need the one or other difference in cmake settings when compared to MSVC. I.e. clang-cl does not like `-Wall` because it triggers excessive warnings, when compared to clang on Linux. Also some intrinsics behave differently on clang-cl than on MSVC. For me, the only change required to get duckdb 0.7.1 building with clang-cl was about one explicit intrinsic definition that MSVC requires but clang-cl does not, so it was easily fixed with one additional ifdef. Here is a diffset, just for the sake of the example:
```
diff --git a/src/include/duckdb/storage/compression/chimp/algorithm/chimp_utils.hpp b/src/include/duckdb/storage/compression/chimp/algorithm/chimp_utils.hpp
index 9ab1438ff8..44ba0d6244 100644
--- a/src/include/duckdb/storage/compression/chimp/algorithm/chimp_utils.hpp
+++ b/src/include/duckdb/storage/compression/chimp/algorithm/chimp_utils.hpp
@@ -15,6 +15,7 @@
 #define __BYTE_ORDER__          __ORDER_LITTLE_ENDIAN__
 #define __ORDER_LITTLE_ENDIAN__ 2
 #include <intrin.h>
+#ifndef __clang__
 static inline int __builtin_ctzll(unsigned long long x) {
 #ifdef _WIN64
 	unsigned long ret;
@@ -70,6 +71,7 @@ static inline int __builtin_clz(unsigned int value) {
 	}
 }
 
+#endif
 #endif
 
 namespace duckdb {
```

But I did not test the resulting library in any way.

So what this shows is that there can be a bit of extra consideration required for using clang-cl. Adding a CI run for it would be a good way to test the support for the compiler, but even better would be to actually use it in development, in case you are developing on Windows?

BTW, Daniel Lemire was just recently writing on why he likes clang-cl: https://lemire.me/blog/2023/02/27/visual-studio-versus-clangcl/
